meadow-client/src/Semantics.purs

```diff
@@ -1159,88 +1154,67 @@ applyTransaction inputs sigs blockNum state contract value = case appResult of
-
-peopleFromContract (When observation timeout contract1 contract2) = S.union (peopleFromObservation observation) (S.union (peopleFromContract contract1) (peopleFromContract contract2))
-
-peopleFromContract (While observation timeout contract1 contract2) = S.union (peopleFromObservation observation) (S.union (peopleFromContract contract1) (peopleFromContract contract2))
-
-peopleFromContract (Scale value1 value2 value3 contract) = S.union (peopleFromValue value1) (S.union (peopleFromValue value2) (S.union (peopleFromValue value3) (peopleFromContract contract)))
-
-peopleFromContract (Let _ contract1 contract2) = S.union (peopleFromContract contract1) (peopleFromContract contract2)
-
-peopleFromContract (Use _) = S.empty
+instance peopleFromValue :: PeopleFrom Value where
+  peopleFrom = cata algebra
+    where
+      algebra :: Algebra ValueF (S.Set Person)
+      algebra CurrentBlockF = S.empty
+      algebra (CommittedF _) = S.empty
+      algebra (ConstantF _) = S.empty
+      algebra (NegValueF value) = value
+      algebra (AddValueF value1 value2) = S.union value1 value2
+      algebra (SubValueF value1 value2) = S.union value1 value2
+      algebra (MulValueF value1 value2) = S.union value1 value2
+      algebra (DivValueF value1 value2 value3) = S.unions [value1, value2, value3]
+      algebra (ModValueF value1 value2 value3) = S.unions [value1, value2, value3]
+      algebra (ValueFromChoiceF v value) = S.insert (unwrap v).person value
+      algebra (ValueFromOracleF _ value) = value
+
+instance peopleFromObservation :: PeopleFrom Observation where
+  peopleFrom = cata algebra
+    where
+      algebra :: Algebra ObservationF (S.Set Person)
+      algebra (BelowTimeoutF _) = S.empty
+      algebra (AndObsF observation1 observation2) = S.union observation1 observation2
+      algebra (OrObsF observation1 observation2) = S.union observation1 observation2
+      algebra (NotObsF observation) = observation
+      algebra (ChoseThisF v _) = S.insert (unwrap v).person S.empty
+      algebra (ChoseSomethingF v) = S.insert (unwrap v).person S.empty
+      algebra (ValueGEF value1 value2) = S.union (peopleFrom value1) (peopleFrom value2)
+      algebra (ValueGTF value1 value2) = S.union (peopleFrom value1) (peopleFrom value2)
+      algebra (ValueLTF value1 value2) = S.union (peopleFrom value1) (peopleFrom value2)
+      algebra (ValueLEF value1 value2) = S.union (peopleFrom value1) (peopleFrom value2)
+      algebra (ValueEQF value1 value2) = S.union (peopleFrom value1) (peopleFrom value2)
+      algebra TrueObsF = S.empty
+      algebra FalseObsF = S.empty
+
+instance peopleFromContract :: PeopleFrom Contract where
+  peopleFrom = cata algebra
+    where
+      algebra :: Algebra ContractF (S.Set Person)
+      algebra NullF = S.empty
+      algebra (CommitF _ _ person value _ _ c1 c2) = S.insert person (S.unions [peopleFrom value, c1, c2])
```

Comment:
Minor, but you could do `... = S.unions [ pure person, peopleFrom value, c1, c2 ]`

Reply:
`Set` isn't a `Functor` :(

Reply:
:-(

Not the first time that's bitten me. `S.singleton` then.

Reply:
There is a way that it _could_ be fixed in Haskell with `ConstraintKinds` but it's not and also I don't think purescript has `ContraintKinds` functionality.